presto-main/src/main/java/com/facebook/presto/sql/rewrite/MaterializedViewOptimizationRewrite.java

```diff
@@ -87,4 +104,58 @@ protected Node visitQuery(Query query, Void context)
             return query;
         }
     }
+
+    private static Query optimizeQueryUsingMaterializedView(
+            Metadata metadata,
+            Session session,
+            SqlParser sqlParser,
+            AccessControl accessControl,
+            Query node)
+    {
+        MaterializedViewCandidateExtractor materializedViewCandidateExtractor = new MaterializedViewCandidateExtractor(session, metadata);
+        materializedViewCandidateExtractor.process(node);
+        Map<QualifiedObjectName, List<QualifiedObjectName>> baseTableToMaterializedViewCandidateMap = materializedViewCandidateExtractor.getMaterializedViewCandidatesForTable();
+        // TODO: Select the most compatible and efficient materialized view for query rewrite optimization https://github.com/prestodb/presto/issues/16431
+        // TODO: Refactor query optimization code https://github.com/prestodb/presto/issues/16759
+
+        Map<QualifiedObjectName, List<ConnectorMaterializedViewDefinition>> baseTableToMaterializedViewDefinitionMap = new HashMap<>();
+        baseTableToMaterializedViewCandidateMap.forEach((baseTable, materializedViewNames) -> {
+            for (QualifiedObjectName materializedViewName : materializedViewNames) {
+                MaterializedViewStatus materializedViewStatus = metadata.getMaterializedViewStatus(session, materializedViewName);
+                // TODO: Refactor this so we increment metric only when rewrite would have occurred
+                if (!(materializedViewStatus.isPartiallyMaterialized() || materializedViewStatus.isFullyMaterialized())
+                        && isMaterializedViewDataConsistencyEnabled(session)) {
+                    session.getRuntimeStats().addMetricValue(MANY_PARTITIONS_MISSING_IN_MATERIALIZED_VIEW_COUNT, NONE, 1);
+                    continue;
+                }
+                ConnectorMaterializedViewDefinition materializedView = metadata.getMaterializedView(session, materializedViewName).orElseThrow(() ->
+                        new IllegalStateException("Materialized view definition not present in metadata as expected."));
+                baseTableToMaterializedViewDefinitionMap.putIfAbsent(baseTable, new ArrayList<>());
+                baseTableToMaterializedViewDefinitionMap.get(baseTable).add(materializedView);
+            }
+        });
+
+        // No fresh materialized views to use - no need to attempt rewrite
+        if (baseTableToMaterializedViewDefinitionMap.isEmpty()) {
+            log.warn("Could not optimize query with materialized view : no materialized views exist which reference base tables: " +
```

Comment:
The error message is misleading, as the materialized view might be present, but not ready to be used. Lets change this to
"Failed to optimize query as no fresh materialized was available for tables: " + {tables}